### DISCUSSION TITLE
How to use PostgreSQL `BIGSERIAL` column type with `.valuesOfRecords(...)`

### DISCUSSION THREAD

**Discussion by foxylion** [jOOQ/jOOQ#14695] [Q&A] [answered]
Hey,
I just found out about JOOQs records and I would like to use them in a query that upserts a lot of rows using `.valuesOfRecords`.

I've created a MCVE to show the problem, this should simplify understanding my problem.
https://github.com/foxylion/jooq-mcve/tree/master/bigserial-primary-key

I've a table that has a primary key using `BIGSERIAL` as data type.
I try to bulk insert multiple records using `.valuesOfRecord(...)`.
I would like the database to auto fill the `id` column with a generated serial form the sequence.

JOOQ is setting the `id` field to `NULL`. This does produce the following SQL error:

```
ERROR: null value in column "id" of relation "segment" violates not-null constraint
```

The generated SQL query looks like the following:

```sql
insert into "mcve"."segment" ("id", "start", "finish", "value", "code")
values (?, ?, ?, ?, ?), (?, ?, ?, ?, ?)
on conflict on constraint "segment_code_key" do update set "id" = excluded."id", "start" = excluded."start", "finish" = excluded."finish", "value" = excluded."value", "code" = excluded."code"
```

I've hoped there is a way to set the field to `DEFAULT` which would delegate the PK generation to PostgreSQL.
Is this possible?

Relevant parts of the code:
- [sql schema](https://github.com/foxylion/jooq-mcve/blob/master/bigserial-primary-key/src/main/resources/db/migration/V1__initialise_database.sql)
- [test case](https://github.com/foxylion/jooq-mcve/blob/master/bigserial-primary-key/src/test/kotlin/org/jooq/mcve/test/kotlin/postgres/KotlinTest.kt#L52)

I appreciate any feedback on this. :-)

**Comment by lukaseder** [comment, marked as answer]
Looks like another occurrence of this bug? https://github.com/jOOQ/jOOQ/issues/14599 It was backported to 3.17.8, but you're still using 3.17.7. Can you confirm?

**Reply by foxylion** [reply]
I just updated the code to the latest version. Issue still persists.
The `onConflict...` related code could just be removed and the issue is still the same.
I kept the `onConflict` part it in my example to show how I will use it later in the real code.

The issue is that JOOQ is setting the `id` field to `NULL` instead of `DEFAULT` when I do not provide any value to the record object.

**Reply by lukaseder** [reply]
OK, thanks for checking. I'll have a closer look at your MCVE tomorrow.

**Reply by lukaseder** [reply]
Ah yes of course, it's not a bug. You specified `*SEGMENT.fields()`, which means you want to insert into the `ID` column as well, but you're not supplying any value to it, so the value is `NULL`. Do this, instead:

```kotlin
ctx().insertInto(SEGMENT)
        .columns(SEGMENT.CODE, SEGMENT.START, SEGMENT.FINISH, SEGMENT.VALUE)
        .valuesOfRecords(data.map { seg ->
            ctx().newRecord(SEGMENT.CODE, SEGMENT.START, SEGMENT.FINISH, SEGMENT.VALUE)
                .values(seg.code, seg.start, seg.finish, seg.value)
        })
        .onConflictOnConstraint(SEGMENT_CODE_KEY)
        .doUpdate()
        .setAllToExcluded()
        .execute()
```

Or even:

```kotlin
ctx().insertInto(SEGMENT)
        .columns(SEGMENT.CODE, SEGMENT.START, SEGMENT.FINISH, SEGMENT.VALUE)
        .valuesOfRows(data.map { seg ->
            DSL.row(seg.code, seg.start, seg.finish, seg.value)
        })
        .onConflictOnConstraint(SEGMENT_CODE_KEY)
        .doUpdate()
        .setAllToExcluded()
        .execute()
```

I understand that there must have been an expectation along the lines that you didn't set the `ID` on the record, so, much like when using `TableRecord.insert()`, you would have expected `DEFAULT` semantics, not `NULL` semantics.

But the difference is that you provided the `ID` column in the `INSERT`'s `columns()` list, so you *must* provide a value. It would be possible to argue that this value should be `DEFAULT`, too, but neither approach is strictly better / more obvious, adn it's not something that we can/should easily change now.

**Reply by foxylion** [reply]
Hi Lukas,
thanks for the detailed explanation, makes sense.
What I personally like about the `ctx().newRecord(SEGMENT).apply { ... }`-style is that it allows me to set fields on a object and the order of assignments does not make any difference. In the second example you provided this is "potentially" error prone.

Example:
```kotlin
ctx().insertInto(SEGMENT)
        .columns(SEGMENT.CODE, SEGMENT.START, SEGMENT.FINISH, SEGMENT.VALUE)
        .valuesOfRows(data.map { seg ->
            DSL.row(seg.code, seg.finish, seg.start, seg.value) // finish and start in wrong order
        })
        .onConflictOnConstraint(SEGMENT_CODE_KEY)
        .doUpdate()
        .setAllToExcluded()
        .execute()
```

As I understand there is currently no API that would allow using records via table definitions.
But exactly for this use case it would be really cool if you could just omit some columns and it would just work without throwing an error that the provided record and the listed columns do not have the same fields. :-)
Maybe something to consider in future JOOQ version as a feature?

**Reply by lukaseder** [reply]
`this.finish = seg.finish` isn't more typesafe. You could have written `this.start = seg.finish` just the same.

**Reply by foxylion** [reply]
This is right, but having ~15 fields, all strings, will make it a little bit more difficult to keep track which field does get which value than checking each line if the assignment is valid.

**Reply by lukaseder** [reply]
Well, since you're using kotlin, you could use `apply` / `let` etc. to call `set()` and `newRecord()` to achieve this style.

In any case, I don't think this is an area worth investing too much. There are already tons of options. More options aren't necessarily going to improve things. It is likely that a new option may address your concern while omitting someone else's.
